Sources/MessagingPush/PushHandling/NotificationCenterFrameworkAdapter.swift

```diff
@@ -0,0 +1,87 @@
+}
+
+/**
+ This class is an adapter that makes our SDK communicate with the iOS framework, `UserNotifications` in production.
+
+ This allows our SDK to not have knowledge of the `UserNotifications` framework, which makes it easier to write automated tests around our SDK.
+
+ Keep this class simple because it is only able to be tested in QA testing. It's meant to be an adapter, not contain logic.
+ */
+// sourcery: InjectRegister = "NotificationCenterFrameworkAdapter"
+class NotificationCenterFrameworkAdapterImpl: NSObject, UNUserNotificationCenterDelegate, NotificationCenterFrameworkAdapter {
+    private let pushEventListener: PushEventListener
+
+    init(pushEventListener: PushEventListener) {
+        self.pushEventListener = pushEventListener
+    }
+
+    var delegate: UNUserNotificationCenterDelegate {
+        self
+    }
+
+    // Functions called by iOS framework, `UserNotifications`. This adapter class simply passes these requests to other code in our SDK where the logic exists.
+
+    func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
+        let wasClickEventHandled = pushEventListener.onPushAction(PushNotification(notification: response.notification), didClickOnPush: response.didClickOnPush)
+
+        if wasClickEventHandled {
+            // call the completion handler so the customer does not need to.
+            completionHandler()
+        }
+    }
+
+    func userNotificationCenter(_ center: UNUserNotificationCenter, willPresent notification: UNNotification, withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
+        let response = pushEventListener.shouldDisplayPushAppInForeground(PushNotification(notification: notification))
+        guard let shouldShowPush = response else {
+            // push not handled by CIO SDK. Exit early. Another click handler will call the completion handler.
+
+            return
+        }
+
+        if shouldShowPush {
+            if #available(iOS 14.0, *) {
+                completionHandler([.list, .banner, .badge, .sound])
+            } else {
+                completionHandler([.badge, .sound])
+            }
+        } else {
+            completionHandler([])
+        }
+    }
+}
+
+// A class that represents a push notification received by the iOS framework, `UserNotifications`.
+// When our SDK receives a push notification from the `UserNotification` framework, the push is converted into
+// an instance of this class, first.
+//
+// This allows us to write automated tests around our SDK's push handling logic because classes inside of `UserUnotifications` internal and not mockable.
+public struct PushNotification {
+    let pushId: String
+    let deliveryDate: Date
+    let title: String
+    let message: String
+    let data: [AnyHashable: Any]
+    let rawNotification: UNNotification
```

Comment:
I am a little confused here, will this be `nullable`? because I thought the reason behind creating this class was (not being able to/have to) create a fake instance of `UNNotification`.

Reply:
I agree, this is confusing. I was hoping that this struct could be a true replacement to `UNNotification`. 

I added it because `CustomerIOParsedPushPayload.swift` depends on `UNNotification` instances. [This line of code](https://github.com/customerio/customerio-ios/pull/447/files#diff-8e556516d6bc8528639b392515340069010e7c0cf0499f60002cc6ef72cfe3acR137) calls `CustomerIOParsedPushPayload.parse`. So we need a `UNNotification` reference to make that `parse()` call. 

I agree, having `UNNotification` in here is not ideal. There are consequences to it. 

🤔 We are using `CustomerIOParsedPushPayload` for 2 different use cases today. Composing rich push notifications as well as parsing a push payload. I am thinking, it could be a value add to split `CustomerIOParsedPushPayload` into 2 separate classes: 1 class that just parses pushes and 1 class that composes a push. 

That could remove the need for `UNNotification`.

Reply:
This PR's focus is on the UserNotification adapter. The change ^^^ will come in a future PR that improves testability of the code. I am glad you asked about this.

Reply:
Glad to hear it's coming in upcoming PRs and I think that change will indeed improve this.